File: past-competitions/2018/src/brett-cpp/Prob06.c
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int main(int argc, char *argv[])
{
	// open file for reading
	FILE *inputFile = fopen("Prob06.in.txt", "r");
	int testCases;

	// read number of test cases
	fscanf(inputFile, "%d\n", &testCases);
	
	char batteryState[7];
	char exchangerState[7];
	char pumpState[7];
	char sensorState[7];
	int overallState, leftState, rightState;

	// execute test cases
	while(testCases > 0){
		// read in status strings - %s covers all text until whitespace is hit
		fscanf(inputFile, "%s %s %s %s\n", batteryState, exchangerState, pumpState, sensorState);

		// reset overall status code
		overallState = 0;
		// check each status and add the appropriate value
		if(strcmp(batteryState, "BROKEN") == 0){
			overallState = overallState + 8;
		}
		if(strcmp(exchangerState, "BROKEN") == 0){
			overallState = overallState + 4;
		}
		if(strcmp(pumpState, "BROKEN") == 0){
			overallState = overallState + 2;
		}
		if(strcmp(sensorState, "BROKEN") == 0){
			overallState = overallState + 1;
		}

		// determine LED states
		// X / 4 is integer division; remainder discarded
		// X % 4 is modular division; remainder is returned
		leftState = overallState / 4;
		rightState = overallState % 4;

		// print LED states
		switch(leftState){
		case 3: printf("blue "); break;
		case 2: printf("green "); break;
		case 1: printf("red "); break;
		case 0: printf("off "); break;
		}
		switch(rightState){
		case 3: printf("blue\n"); break;
		case 2: printf("green\n"); break;
		case 1: printf("red\n"); break;
		case 0: printf("off\n"); break;
		}

		testCases = testCases - 1;
	}

	// close the file
	fclose(inputFile);
}

File: past-competitions/2018/src/brett-python/Prob06.py
# Open the input file
with open("Prob06.in.txt", "rt") as inputFile:
    # Read the number of test cases (trim out the newline)
    cases = int(inputFile.readline().replace("\n", ""))
    # For each test case
    for caseNum in range(cases):
        # Read the line
        line = inputFile.readline().replace("\n", "")
        # Break up words
        statuses = line.split(" ")
        value = 8
        errorCode = 0
        # Check each status
        for status in statuses:
            if status == "BROKEN":
                # If broken, add the value to the error code
                errorCode = errorCode + value
            # Reduce the value by half for each component
            value = value / 2
        left = "off"
        right = "off"
        # integer division determines left light
        leftValue = errorCode // 4
        # modulo division determines right light
        rightValue = errorCode % 4
        if leftValue == 1:
            left = "red"
        elif leftValue == 2:
            left = "green"
        elif leftValue == 3:
            left = "blue"
        if rightValue == 1:
            right = "red"
        elif rightValue == 2:
            right = "green"
        elif rightValue == 3:
            right = "blue"
        # print light colors
        print(left + " " + right)
